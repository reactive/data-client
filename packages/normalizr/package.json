{
  "name": "@data-client/normalizr",
  "version": "0.14.17",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://dataclient.io/docs/concepts/normalization",
  "keywords": [
    "normalize",
    "denormalize",
    "flux",
    "redux",
    "data",
    "model",
    "cache",
    "api",
    "memoize",
    "react-native",
    "ios",
    "android",
    "web",
    "schema",
    "typescript",
    "validation",
    "relational",
    "reactive",
    "json"
  ],
  "sideEffects": false,
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.4": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.1": {
      "": [
        "ts4.1/index.d.ts"
      ],
      "*": [
        "ts4.1/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "node.mjs",
    "legacy",
    "ts4.1",
    "ts4.0",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development yarn g:rollup",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production yarn g:rollup",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean ts3.4 ts4.0 ts4.1",
    "build:legacy-types": "yarn g:legacy-types 4.1 4.0 3.4",
    "prepublishOnly": "run build"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/normalizr"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "core-js-pure": "^3.40.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/endpoint": "workspace:*",
    "@js-temporal/polyfill": "^0.4.4",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "immutable": "4.2.2",
    "rollup-plugins": "workspace:*"
  }
}
