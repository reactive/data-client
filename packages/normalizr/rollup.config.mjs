import {
  babel,
  commonjs,
  filesize,
  dts,
  resolve,
  terser,
  onwarn,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies).filter(
  dep => !['@babel/runtime'].includes(dep),
);

const name = 'normalizr';

const isProduction = process.env.NODE_ENV === 'production';

const destBase = 'dist/normalizr';
const destExtension = `${isProduction ? '.min' : ''}.js`;
const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];

function snakeCase(id) {
  return id.replace(/(-|\/)/g, '_').replace(/@/g, '');
}

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];

if (process.env.BROWSERSLIST_ENV !== 'node12') {
  if (process.env.NODE_ENV === 'production') {
    configs.push({
      input: 'src/index.ts',
      output: [{ file: `${destBase}.es${destExtension}`, format: 'es' }],
      external: id => id === '..' || isExternal(id),
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
  } else {
    configs.push({
      input: 'src/index.ts',
      output: [
        {
          file: `${destBase}.umd${destExtension}`,
          format: 'umd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.amd${destExtension}`,
          format: 'amd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.browser${destExtension}`,
          format: 'iife',
          name: snakeCase(name),
        },
      ],
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
          caller: { polyfillMethod: false },
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
    configs.push({
      input: './lib/index.d.ts',
      output: [{ file: 'index.d.ts', format: 'es' }],
      plugins: [dts()],
    });
  }
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'src/index.ts',
    output: [{ file: `${destBase}${destExtension}`, format: 'cjs' }],
    external: id => id === '..' || isExternal(id),
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        babelHelpers: 'runtime',
        rootMode: 'upward',
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      isProduction && terser(),
      filesize(),
    ].filter(Boolean),
  });
}
export default configs;
