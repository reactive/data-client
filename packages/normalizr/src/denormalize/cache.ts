import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
