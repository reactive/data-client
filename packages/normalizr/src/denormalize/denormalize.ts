import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
