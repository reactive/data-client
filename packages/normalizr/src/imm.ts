export { MemoPolicy } from './memo/Policy.imm.js';
export { denormalize } from './denormalize/denormalize.imm.js';
