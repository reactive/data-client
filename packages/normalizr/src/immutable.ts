export { DelegateImmutable } from './memo/Delegate.immutable.js';
