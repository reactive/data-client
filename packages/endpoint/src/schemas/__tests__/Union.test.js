// eslint-env jest
import { normalize } from '@data-client/normalizr';
import { denormalize as plainDenormalize } from '@data-client/normalizr';
import { denormalize as immDenormalize } from '@data-client/normalizr/imm';
import { IDEntity } from '__tests__/new';
import { waterfallSchema } from '__tests__/UnionSchema';
import { fromJS } from 'immutable';

import { SimpleMemoCache, fromJSEntities } from './denormalize';
import { schema } from '../../';

let dateSpy;
beforeAll(() => {
  dateSpy = jest

    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});
let warnSpy;
afterEach(() => {
  warnSpy.mockRestore();
});
beforeEach(() => {
  (warnSpy = jest.spyOn(console, 'warn')).mockImplementation(() => {});
});

describe(`${schema.Union.name} normalization`, () => {
  test('throws if not given a schemaAttribute', () => {
    expect(() => new schema.Union({})).toThrow();
  });

  test('normalizes an object using string schemaAttribute', () => {
    class User extends IDEntity {}
    class Group extends IDEntity {}
    const union = new schema.Union(
      {
        users: User,
        groups: Group,
      },
      'type',
    );

    expect(normalize(union, { id: '1', type: 'users' })).toMatchSnapshot();
    expect(normalize(union, { id: '2', type: 'groups' })).toMatchSnapshot();
  });

  test('normalizes an array of multiple entities using a function to infer the schemaAttribute', () => {
    class User extends IDEntity {}
    class Group extends IDEntity {}
    const union = new schema.Union(
      {
        users: User,
        groups: Group,
      },
      input => {
        return (
          input.username ? 'users'
          : input.groupname ? 'groups'
          : null
        );
      },
    );

    expect(normalize(union, { id: '1', username: 'Janey' })).toMatchSnapshot();
    expect(
      normalize(union, { id: '2', groupname: 'People' }),
    ).toMatchSnapshot();
    expect(normalize(union, { id: '3', notdefined: 'yep' })).toMatchSnapshot();
    expect(warnSpy.mock.calls).toMatchSnapshot();
  });
});

describe(`${schema.Union.name} buildQueryKey`, () => {
  class User extends IDEntity {
    static key = 'User';
  }
  class Group extends IDEntity {
    static key = 'Group';
  }

  // Common schema definitions
  const stringAttributeUnion = new schema.Union(
    {
      users: User,
      groups: Group,
    },
    'type',
  );

  const functionAttributeUnion = new schema.Union(
    {
      users: User,
      groups: Group,
    },
    input => {
      return (
        input.username ? 'users'
        : input.groupname ? 'groups'
        : undefined
      );
    },
  );

  test('buildQueryKey with discriminator in args', () => {
    const memo = new SimpleMemoCache();

    const state = {
      entities: {
        User: {
          1: { id: '1', username: 'Janey', type: 'users' },
        },
        Group: {
          2: { id: '2', groupname: 'People', type: 'groups' },
        },
      },
      indexes: {},
    };

    // Fast case - args include type discriminator
    const result1 = memo.memo.buildQueryKey(
      stringAttributeUnion,
      [{ id: '1', type: 'users' }],
      state,
    );
    expect(result1).toEqual({ id: '1', schema: 'users' });

    const result2 = memo.memo.buildQueryKey(
      stringAttributeUnion,
      [{ id: '2', type: 'groups' }],
      state,
    );
    expect(result2).toEqual({ id: '2', schema: 'groups' });
  });

  test('buildQueryKey without discriminator - fallback case', () => {
    const memo = new SimpleMemoCache();

    const state = {
      entities: {
        User: {
          1: { id: '1', username: 'Janey', type: 'users' },
        },
        Group: {
          2: { id: '2', groupname: 'People', type: 'groups' },
        },
      },
      indexes: {},
    };

    // Fallback case - args missing type discriminator, only {id}
    // Should try every possible schema until it finds a match
    const result1 = memo.memo.buildQueryKey(
      stringAttributeUnion,
      [{ id: '1' }],
      state,
    );
    expect(result1).toEqual({ id: '1', schema: 'users' });

    const result2 = memo.memo.buildQueryKey(
      stringAttributeUnion,
      [{ id: '2' }],
      state,
    );
    expect(result2).toEqual({ id: '2', schema: 'groups' });
  });

  test('buildQueryKey with function schemaAttribute missing discriminator', () => {
    const memo = new SimpleMemoCache();

    const state = {
      entities: {
        User: {
          1: { id: '1', username: 'Janey' },
        },
        Group: {
          2: { id: '2', groupname: 'People' },
        },
      },
      indexes: {},
    };

    // With function schemaAttribute, args missing username/groupname
    // Should fallback to trying every schema
    const result1 = memo.memo.buildQueryKey(
      functionAttributeUnion,
      [{ id: '1' }],
      state,
    );
    expect(result1).toEqual({ id: '1', schema: 'users' });

    const result2 = memo.memo.buildQueryKey(
      functionAttributeUnion,
      [{ id: '2' }],
      state,
    );
    expect(result2).toEqual({ id: '2', schema: 'groups' });
  });

  test('buildQueryKey returns undefined when no entity found', () => {
    const memo = new SimpleMemoCache();

    const state = {
      entities: {
        User: {},
        Group: {},
      },
      indexes: {},
    };

    // No entity exists with id '999'
    const result = memo.memo.buildQueryKey(
      stringAttributeUnion,
      [{ id: '999' }],
      state,
    );
    expect(result).toBeUndefined();
  });

  test('buildQueryKey returns undefined when no args provided', () => {
    const memo = new SimpleMemoCache();

    const state = {
      entities: {},
      indexes: {},
    };

    // No args provided
    const result = memo.memo.buildQueryKey(stringAttributeUnion, [], state);
    expect(result).toBeUndefined();
  });
});

describe('complex case', () => {
  test('works with undefined', () => {
    const response = {
      sequences: [
        null,
        undefined,
        {
          group_a_stats: {
            count: 79557,
            sum: 3727219.2099861577,
          },
          group_b_stats: {
            count: 66794,
            sum: 2999496.370384956,
          },
          sequence_id: 22,
          sequence_type: 'temporal_mean_shift',
          subpops_group_a_incremental_stats: [
            {
              count: 19302,
              sum: 909835.9502538516,
            },
            {
              count: 12452,
              sum: 468902.2598208785,
            },
            {
              count: 29674,
              sum: 1533126.7700431347,
            },
            {
              count: 3979,
              sum: 123931.08003079893,
            },
            {
              count: 3501,
              sum: 173706.13005542755,
            },
          ],
          subpops_group_a_stats: [
            {
              count: 19302,
              sum: 909835.9502538516,
            },
            {
              count: 15550,
              sum: 656345.8200406507,
            },
            {
              count: 38523,
              sum: 1767670.8099250195,
            },
            {
              count: 15778,
              sum: 549563.3501347005,
            },
            {
              count: 20194,
              sum: 951307.5302157328,
            },
          ],
          subpops_group_b_incremental_stats: [
            {
              count: 16022,
              sum: 686623.8400082085,
            },
            {
              count: 10427,
              sum: 632778.0701571181,
            },
            {
              count: 25270,
              sum: 1002223.6299012974,
            },
            {
              count: 3386,
              sum: 104987.2500332594,
            },
            {
              count: 2893,
              sum: 139013.64997577667,
            },
          ],
          subpops_group_b_stats: [
            {
              count: 16022,
              sum: 686623.8400082085,
            },
            {
              count: 12964,
              sum: 713309.7901891638,
            },
            {
              count: 32773,
              sum: 1542281.990038909,
            },
            {
              count: 12722,
              sum: 423195.7500065565,
            },
            {
              count: 17110,
              sum: 768182.7200815715,
            },
          ],
          subpopulations: [
            [
              {
                absolute_range: null,
                factor_type: 'value',
                keyword: null,
                percentile_range: null,
                value: {
                  column: 'loyalty_program',
                  data_type: 'string',
                  value: 'none',
                },
              },
            ],
            [
              {
                absolute_range: null,
                factor_type: 'value',
                keyword: null,
                percentile_range: null,
                value: {
                  column: 'coupon_code',
                  data_type: 'string',
                  value: '01DEC2019PROMO',
                },
              },
            ],
            [
              {
                absolute_range: null,
                factor_type: 'value',
                keyword: null,
                percentile_range: null,
                value: {
                  column: 'loyalty_program',
                  data_type: 'string',
                  value: 'silver',
                },
              },
            ],
            [
              {
                absolute_range: null,
                factor_type: 'value',
                keyword: null,
                percentile_range: null,
                value: {
                  column: 'ad_shown',
                  data_type: 'string',
                  value: 'week52_20',
                },
              },
            ],
            [
              {
                absolute_range: null,
                factor_type: 'value',
                keyword: null,
                percentile_range: null,
                value: {
                  column: 'day_part',
                  data_type: 'string',
                  value: 'afternoon',
                },
              },
            ],
          ],
        },
      ],
    };
    const denorm = normalize(waterfallSchema, response);
    expect(denorm).toMatchSnapshot();
    expect(
      new SimpleMemoCache().denormalize(
        waterfallSchema,
        denorm.result,
        denorm.entities,
      ),
    ).toMatchSnapshot();
  });
});

class User extends IDEntity {}
class Group extends IDEntity {}
const entities = {
  User: {
    1: { id: '1', username: 'Janey', type: 'users' },
  },
  Group: {
    2: { id: '2', groupname: 'People', type: 'groups' },
  },
};
describe.each([
  ['direct', plainDenormalize, data => data, data => data],
  ['immutable', immDenormalize, fromJS, fromJSEntities],
])(`input (%s)`, (_, denormalize, createInput, createEntities) => {
  describe.each([['current', denormalize]])(
    `${schema.Union.name} denormalization (%s)`,
    (_, denormalize) => {
      test('denormalizes an object using string schemaAttribute', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          'type',
        );

        expect(
          denormalize(
            union,
            createInput({ id: '1', schema: 'users' }),
            createEntities(entities),
          ),
        ).toMatchSnapshot();

        expect(
          denormalize(
            union,
            createInput({ id: '2', schema: 'groups' }),
            createEntities(entities),
          ),
        ).toMatchSnapshot();
      });

      test('denormalizes an array of multiple entities using a function to infer the schemaAttribute', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          input => {
            return input.username ? 'users' : 'groups';
          },
        );

        expect(
          denormalize(
            union,
            createInput({ id: '1', schema: 'users' }),
            createEntities(entities),
          ),
        ).toMatchSnapshot();

        expect(
          denormalize(
            union,
            createInput({ id: '2', schema: 'groups' }),
            createEntities(entities),
          ),
        ).toMatchSnapshot();
      });

      test('returns the original value when no schema is given', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          input => {
            return (
              input.username ? 'users'
              : input.groupname ? 'groups'
              : undefined
            );
          },
        );

        expect(
          denormalize(
            union,
            createInput({ id: '1' }),
            createEntities(entities),
          ),
        ).toMatchSnapshot();
        expect(warnSpy.mock.calls).toMatchSnapshot();
      });

      test('returns the original value when string is given', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          input => {
            return (
              input.username ? 'users'
              : input.groupname ? 'groups'
              : undefined
            );
          },
        );

        expect(
          denormalize(union, '1', createEntities(entities)),
        ).toMatchSnapshot();
        expect(warnSpy.mock.calls).toMatchSnapshot();
      });

      test('returns the original value when null is given', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          input => {
            return (
              input.username ? 'users'
              : input.groupname ? 'groups'
              : undefined
            );
          },
        );

        expect(denormalize(union, null, createEntities(entities))).toBeNull();
      });

      test('returns the original value when undefined is given', () => {
        const union = new schema.Union(
          {
            users: User,
            groups: Group,
          },
          input => {
            return (
              input.username ? 'users'
              : input.groupname ? 'groups'
              : undefined
            );
          },
        );

        expect(
          denormalize(union, undefined, createEntities(entities)),
        ).toBeUndefined();
      });
    },
  );
});
