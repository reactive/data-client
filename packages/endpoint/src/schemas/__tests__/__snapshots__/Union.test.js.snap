// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
        "username": "Janey",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "id": "1",
    "schema": "users",
  },
}
`;

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 2`] = `
{
  "entities": {
    "Group": {
      "2": {
        "groupname": "People",
        "id": "2",
      },
    },
  },
  "entitiesMeta": {
    "Group": {
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "id": "2",
    "schema": "groups",
  },
}
`;

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 3`] = `
{
  "entities": {},
  "entitiesMeta": {},
  "indexes": {},
  "result": {
    "id": "3",
    "notdefined": "yep",
  },
}
`;

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 4`] = `
[
  [
    "Schema attribute null is not expected.
Expected one of: "users", "groups"

Value: {
  "id": "3",
  "notdefined": "yep"
}",
  ],
]
`;

exports[`UnionSchema normalization normalizes an object using string schemaAttribute 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
        "type": "users",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "id": "1",
    "schema": "users",
  },
}
`;

exports[`UnionSchema normalization normalizes an object using string schemaAttribute 2`] = `
{
  "entities": {
    "Group": {
      "2": {
        "id": "2",
        "type": "groups",
      },
    },
  },
  "entitiesMeta": {
    "Group": {
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "id": "2",
    "schema": "groups",
  },
}
`;

exports[`complex case works with undefined 1`] = `
{
  "entities": {
    "ContinuousSequence": {
      "22": {
        "group_a_stats": {
          "count": 79557,
          "sum": 3727219.2099861577,
        },
        "group_b_stats": {
          "count": 66794,
          "sum": 2999496.370384956,
        },
        "sequence_id": 22,
        "sequence_type": "temporal_mean_shift",
        "subpops_group_a_incremental_stats": [
          {
            "count": 19302,
            "sum": 909835.9502538516,
          },
          {
            "count": 12452,
            "sum": 468902.2598208785,
          },
          {
            "count": 29674,
            "sum": 1533126.7700431347,
          },
          {
            "count": 3979,
            "sum": 123931.08003079893,
          },
          {
            "count": 3501,
            "sum": 173706.13005542755,
          },
        ],
        "subpops_group_a_stats": [
          {
            "count": 19302,
            "sum": 909835.9502538516,
          },
          {
            "count": 15550,
            "sum": 656345.8200406507,
          },
          {
            "count": 38523,
            "sum": 1767670.8099250195,
          },
          {
            "count": 15778,
            "sum": 549563.3501347005,
          },
          {
            "count": 20194,
            "sum": 951307.5302157328,
          },
        ],
        "subpops_group_b_incremental_stats": [
          {
            "count": 16022,
            "sum": 686623.8400082085,
          },
          {
            "count": 10427,
            "sum": 632778.0701571181,
          },
          {
            "count": 25270,
            "sum": 1002223.6299012974,
          },
          {
            "count": 3386,
            "sum": 104987.2500332594,
          },
          {
            "count": 2893,
            "sum": 139013.64997577667,
          },
        ],
        "subpops_group_b_stats": [
          {
            "count": 16022,
            "sum": 686623.8400082085,
          },
          {
            "count": 12964,
            "sum": 713309.7901891638,
          },
          {
            "count": 32773,
            "sum": 1542281.990038909,
          },
          {
            "count": 12722,
            "sum": 423195.7500065565,
          },
          {
            "count": 17110,
            "sum": 768182.7200815715,
          },
        ],
        "subpopulations": [
          [
            {
              "absolute_range": null,
              "factor_type": "value",
              "keyword": null,
              "percentile_range": null,
              "value": {
                "column": "loyalty_program",
                "data_type": "string",
                "value": "none",
              },
            },
          ],
          [
            {
              "absolute_range": null,
              "factor_type": "value",
              "keyword": null,
              "percentile_range": null,
              "value": {
                "column": "coupon_code",
                "data_type": "string",
                "value": "01DEC2019PROMO",
              },
            },
          ],
          [
            {
              "absolute_range": null,
              "factor_type": "value",
              "keyword": null,
              "percentile_range": null,
              "value": {
                "column": "loyalty_program",
                "data_type": "string",
                "value": "silver",
              },
            },
          ],
          [
            {
              "absolute_range": null,
              "factor_type": "value",
              "keyword": null,
              "percentile_range": null,
              "value": {
                "column": "ad_shown",
                "data_type": "string",
                "value": "week52_20",
              },
            },
          ],
          [
            {
              "absolute_range": null,
              "factor_type": "value",
              "keyword": null,
              "percentile_range": null,
              "value": {
                "column": "day_part",
                "data_type": "string",
                "value": "afternoon",
              },
            },
          ],
        ],
      },
    },
  },
  "entitiesMeta": {
    "ContinuousSequence": {
      "22": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "sequences": [
      null,
      undefined,
      {
        "id": "22",
        "schema": "temporal_mean_shift",
      },
    ],
  },
}
`;

exports[`complex case works with undefined 2`] = `
{
  "sequences": [
    null,
    ContinuousSequence {
      "group_a_stats": {
        "count": 79557,
        "sum": 3727219.2099861577,
      },
      "group_b_stats": {
        "count": 66794,
        "sum": 2999496.370384956,
      },
      "sequence_id": 22,
      "sequence_type": "temporal_mean_shift",
      "subpops_group_a_incremental_stats": [
        {
          "count": 19302,
          "sum": 909835.9502538516,
        },
        {
          "count": 12452,
          "sum": 468902.2598208785,
        },
        {
          "count": 29674,
          "sum": 1533126.7700431347,
        },
        {
          "count": 3979,
          "sum": 123931.08003079893,
        },
        {
          "count": 3501,
          "sum": 173706.13005542755,
        },
      ],
      "subpops_group_a_stats": [
        {
          "count": 19302,
          "sum": 909835.9502538516,
        },
        {
          "count": 15550,
          "sum": 656345.8200406507,
        },
        {
          "count": 38523,
          "sum": 1767670.8099250195,
        },
        {
          "count": 15778,
          "sum": 549563.3501347005,
        },
        {
          "count": 20194,
          "sum": 951307.5302157328,
        },
      ],
      "subpops_group_b_incremental_stats": [
        {
          "count": 16022,
          "sum": 686623.8400082085,
        },
        {
          "count": 10427,
          "sum": 632778.0701571181,
        },
        {
          "count": 25270,
          "sum": 1002223.6299012974,
        },
        {
          "count": 3386,
          "sum": 104987.2500332594,
        },
        {
          "count": 2893,
          "sum": 139013.64997577667,
        },
      ],
      "subpops_group_b_stats": [
        {
          "count": 16022,
          "sum": 686623.8400082085,
        },
        {
          "count": 12964,
          "sum": 713309.7901891638,
        },
        {
          "count": 32773,
          "sum": 1542281.990038909,
        },
        {
          "count": 12722,
          "sum": 423195.7500065565,
        },
        {
          "count": 17110,
          "sum": 768182.7200815715,
        },
      ],
      "subpopulations": [
        [
          {
            "absolute_range": null,
            "factor_type": "value",
            "keyword": null,
            "percentile_range": null,
            "value": {
              "column": "loyalty_program",
              "data_type": "string",
              "value": "none",
            },
          },
        ],
        [
          {
            "absolute_range": null,
            "factor_type": "value",
            "keyword": null,
            "percentile_range": null,
            "value": {
              "column": "coupon_code",
              "data_type": "string",
              "value": "01DEC2019PROMO",
            },
          },
        ],
        [
          {
            "absolute_range": null,
            "factor_type": "value",
            "keyword": null,
            "percentile_range": null,
            "value": {
              "column": "loyalty_program",
              "data_type": "string",
              "value": "silver",
            },
          },
        ],
        [
          {
            "absolute_range": null,
            "factor_type": "value",
            "keyword": null,
            "percentile_range": null,
            "value": {
              "column": "ad_shown",
              "data_type": "string",
              "value": "week52_20",
            },
          },
        ],
        [
          {
            "absolute_range": null,
            "factor_type": "value",
            "keyword": null,
            "percentile_range": null,
            "value": {
              "column": "day_part",
              "data_type": "string",
              "value": "afternoon",
            },
          },
        ],
      ],
    },
  ],
}
`;

exports[`input (direct) UnionSchema denormalization (current) denormalizes an array of multiple entities using a function to infer the schemaAttribute 1`] = `
User {
  "id": "1",
  "type": "users",
  "username": "Janey",
}
`;

exports[`input (direct) UnionSchema denormalization (current) denormalizes an array of multiple entities using a function to infer the schemaAttribute 2`] = `
Group {
  "groupname": "People",
  "id": "2",
  "type": "groups",
}
`;

exports[`input (direct) UnionSchema denormalization (current) denormalizes an object using string schemaAttribute 1`] = `
User {
  "id": "1",
  "type": "users",
  "username": "Janey",
}
`;

exports[`input (direct) UnionSchema denormalization (current) denormalizes an object using string schemaAttribute 2`] = `
Group {
  "groupname": "People",
  "id": "2",
  "type": "groups",
}
`;

exports[`input (direct) UnionSchema denormalization (current) returns the original value when no schema is given 1`] = `
{
  "id": "1",
}
`;

exports[`input (direct) UnionSchema denormalization (current) returns the original value when no schema is given 2`] = `
[
  [
    "TypeError: Unable to infer schema for UnionSchema
Value: {
  "id": "1"
}.",
  ],
]
`;

exports[`input (direct) UnionSchema denormalization (current) returns the original value when string is given 1`] = `"1"`;

exports[`input (direct) UnionSchema denormalization (current) returns the original value when string is given 2`] = `
[
  [
    "TypeError: Unable to infer schema for UnionSchema
Value: "1".",
  ],
]
`;

exports[`input (immutable) UnionSchema denormalization (current) denormalizes an array of multiple entities using a function to infer the schemaAttribute 1`] = `
User {
  "id": "1",
  "type": "users",
  "username": "Janey",
}
`;

exports[`input (immutable) UnionSchema denormalization (current) denormalizes an array of multiple entities using a function to infer the schemaAttribute 2`] = `
Group {
  "groupname": "People",
  "id": "2",
  "type": "groups",
}
`;

exports[`input (immutable) UnionSchema denormalization (current) denormalizes an object using string schemaAttribute 1`] = `
User {
  "id": "1",
  "type": "users",
  "username": "Janey",
}
`;

exports[`input (immutable) UnionSchema denormalization (current) denormalizes an object using string schemaAttribute 2`] = `
Group {
  "groupname": "People",
  "id": "2",
  "type": "groups",
}
`;

exports[`input (immutable) UnionSchema denormalization (current) returns the original value when no schema is given 1`] = `
Immutable.Map {
  "id": "1",
}
`;

exports[`input (immutable) UnionSchema denormalization (current) returns the original value when no schema is given 2`] = `
[
  [
    "TypeError: Unable to infer schema for UnionSchema
Value: {
  "id": "1"
}.",
  ],
]
`;

exports[`input (immutable) UnionSchema denormalization (current) returns the original value when string is given 1`] = `"1"`;

exports[`input (immutable) UnionSchema denormalization (current) returns the original value when string is given 2`] = `
[
  [
    "TypeError: Unable to infer schema for UnionSchema
Value: "1".",
  ],
]
`;
